google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/BlockingPullSubscriberImpl.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.internal;
+
+import com.google.api.core.ApiService.Listener;
+import com.google.api.core.ApiService.State;
+import com.google.cloud.pubsublite.SequencedMessage;
+import com.google.cloud.pubsublite.cloudpubsub.FlowControlSettings;
+import com.google.cloud.pubsublite.internal.wire.Subscriber;
+import com.google.cloud.pubsublite.internal.wire.SubscriberFactory;
+import com.google.cloud.pubsublite.proto.FlowControlRequest;
+import com.google.cloud.pubsublite.proto.SeekRequest;
+import com.google.common.util.concurrent.Monitor;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class BlockingPullSubscriberImpl implements BlockingPullSubscriber {
+
+  private final Subscriber underlying;
+
+  @GuardedBy("monitor.monitor")
+  private Optional<CheckedApiException> error = Optional.empty();
+
+  @GuardedBy("monitor.monitor")
+  private Deque<SequencedMessage> messages = new ArrayDeque<>();
+
+  private final CloseableMonitor monitor = new CloseableMonitor();
```

Comment:
Don't implement as "watch" since that requires a new thread pool. Implement as "set when new messages if exists". In other words:

```java
@GuardedBy("this")
private Optional<SettableApiFuture<Void>> notification = Optional.empty();
...
@Override
public synchronized ApiFuture<Void> onData() {
  if (error.isPresent()) return ApiFutures.immediateFailedFuture(error.get());
  if (!messages.empty()) return ApiFutures.immediateFuture(null);
  if (!notification.isPresent()) notification = Optional.of(SettableApiFuture.create());
  return notification.get();
}

@Override
public synchronized void addMessages(Collection<SequencedMessage> new_messages) {
  if (notification.isPresent()) {
    notification.get().set(null);
    notification = Optional.empty();
  }
  ...
}
```

and fail it when underlying fails.